### PR TITLE
Changed Be/NotBe on IComparable to use Object.Equals and added BeRankedEquallyTo

### PR DESCRIPTION
**Pull request by rubenrorije**
A new pull request for #172 in which all commits have been cleaned up. I struggled with getting Git to behave the way I wanted, but I think this should be it. 

Just to reiterate, this might be a breaking change when released since it does not check CompareTo == 0 anymore. This might be something people depend upon.

### CODE REVIEW

**Comment by jnyrup** [comment]
I agree on the change and think the potential runtime breakage sounds reasonable.

From [`IComparer<T>`](https://docs.microsoft.com/en-us/dotnet/api/system.icomparable-1?view=netframework-4.8)
> Defines a generalized comparison method that a value type or class implements to create a type-specific comparison method for ordering or sorting its instances.

Can we offer an alternative or guidance for users who still want to test that two items are equal using their `CompareTo` method?
Ugly workarounds would be
```c#
subject.Should().BeLessOrEqualTo(expected)
    .And.BeGreaterOrEqualTo(expected);

subject.Should().BeInRange(expected, expected);
```

Explicit test but inferior failure message.
```c#
subject.CompareTo(expected).Should().Be(0);
```

**Comment by rubenrorije** [comment]
How about naming it `subject.Should().BeRankedEqualTo(expected)`?
About guidance, when such a method exists an explanation and reference can be included in the comments of `Be` and `NotBe`.

File: Tests/Shared.Specs/ComparableSpecs.cs [hunk 1]
```diff
@@ -538,6 +571,43 @@ public override string ToString()
         }
     }
 
+    public class SameInstanceIsNotEqualClass
```
Comment (dennisdoomen) [inline Tests/Shared.Specs/ComparableSpecs.cs]: I would keep these helper classes right after the last test method that uses it. We did this differently in the past, but I think this aligns closer to the "reads like a book" from Clean Code.

**Comment by dennisdoomen** [comment]
> How about naming it `subject.Should().BeRankedEqualTo(expected)`?

Love that one. Probably should be `BeRankedAsEqualTo`

**Comment by dennisdoomen** [comment]
@rubenrorije are you planning to add that `BeRankedAsEqualTo`?

**Comment by rubenrorije** [comment]
Sorry, made the same git mistakes again. I am obviously still a noob on the git part. I will try to fix this.

**Comment by rubenrorije** [comment]
Could you give me some pointers to resolve the git issues here? I assume that the rebase I did resulted in the commits 9050e37 and c477f8b to be included. But I am a bit lost how to resolve this.

**Comment by jnyrup** [comment]
@rubenrorije It looks like you pulled `master` after rebasing on it.
When rebasing, you'll need to push force.
`git push --force`/`git push -f` (There's also a `force-with-lease` option that does some additional checking)

In git commits are identified by a hash value computed on the difference between the commit and its parent commit.

Before rebasing both your local branch and the remote branch on github looks like:
```
master --> ... --> HEAD
 |
 + --> PR1177
```

After rebasing your local branch the picture is now
local:
```
master --> ... --> HEAD
                    |
                    + --> PR1177' // note the '
```

remote:
```
master --> ... --> HEAD
 |
 + --> PR1177
```

As `PR1177` and `PR1177'` has different parent commits locally and remote, git sees them as two different commits.
This step can be confusing as git will now report that there are remote changes, which your local branch does not have. 
It's technically true, due to different hash values, but we *know* it's because of our rebase.

So if you pull master instead of force pushing the local picture becomes:
```
master --> ... --> HEAD
                    |
                    + --> PR1177 --> PR1177'
```

If you rebase again on `master`, you can remove the unwanted commits from the history of this branch.
I usually do this via terminal
```
git rebase -i master
```
The terminal opens some editor (for me it's Vim)
I then write replace `pick` with `drop` for the commits I want to remove.
Save the file and exit.
The rewritten history can be seen with
`git log master...`
Then `git push -f` to push the local branch to the remote repository.

**Comment by rubenrorije** [comment]
Thanks for the extensive explanation. That made things much clearer. I was able to fix the issues now. I did not want to screw everything up again by rebasing again. When the pull request is not accepted directly it must be done again anyway.

**Comment by jnyrup** [comment]
I haven't fallen in love with `BeRankedAsEqualTo` (yet), so I would like to brainstorm a bit on the naming before settling.

* `BeComparableTo`
* `CompareEqualTo`
* `RankEqualTo`
* `BeEqualTo`

`BeEqualTo` is not used anywhere else and it's in line with naming of e.g. `BeGreaterThan`, but I'm not sure whether it's confusingly close to `Be`

**Comment by rubenrorije** [comment]
I agree discussing this a bit more might be useful. I do not have a strong preference. 

> `BeEqualTo` is not used anywhere else and it's in line with naming of e.g. `BeGreaterThan`, but I'm not sure whether it's confusingly close to `Be`

If I would see the options `Be` and `BeEqualTo` after typing `Should()` I would (wrongfully) assume from the method names that `Be` checks reference equality and `BeEqualTo` equality. When also considering the similarly named `BeEquivalentTo`, it might not be that intuitive which one does what.

Slightly related and maybe against the whole idea of the project, but I think it would be nice to have one (additional?) function to check all comparisons:
`subject.Should().BeComparable(instanceLess, instanceEqual, instanceGreater, because)` that combines all checks.  I assume this is done now by multiple test methods and/or multiple assertions within a test method.

**Comment by dennisdoomen** [comment]
I could live with `BeComparableTo`, but not with the others.

**Comment by dennisdoomen** [comment]
What about `BeRankedEquallyTo`?

**Comment by jnyrup** [comment]
I'm fine with `BeRankedEquallyTo`.

**Comment by rubenrorije** [comment]
I do not have any preference, so I renamed the methods.

File: Src/FluentAssertions/Numeric/ComparableTypeAssertions.cs [hunk 2]
```diff
@@ -122,12 +171,12 @@ public AndConstraint<ComparableTypeAssertions<T>> Be(T expected, string because
         /// <param name="becauseArgs">
         /// Zero or more objects to format using the placeholders in <see cref="because"/>.
         /// </param>
-        public AndConstraint<ComparableTypeAssertions<T>> NotBe(T unexpected, string because = "", params object[] becauseArgs)
+        public AndConstraint<ComparableTypeAssertions<T>> NotBeRankedEquallyTo(T expected, string because = "", params object[] becauseArgs)
```
Comment (jnyrup) [inline Src/FluentAssertions/Numeric/ComparableTypeAssertions.cs]: `expected` --> `unexpected`

**Comment by dennisdoomen** [comment]
What's stopping us from merging this one?

File: Src/FluentAssertions/Numeric/ComparableTypeAssertions.cs [hunk 3]
```diff
@@ -110,7 +111,55 @@ public AndConstraint<ComparableTypeAssertions<T>> Be(T expected, string because
...
+        /// </param>
+        /// <param name="becauseArgs">
+        /// Zero or more values to use for filling in any <see cref="string.Format(string,object[])" /> compatible placeholders.
+        /// </param>
+        public AndConstraint<ComparableTypeAssertions<T>> BeRankedEquallyTo(T expected, string because = "", params object[] becauseArgs)
+        {
+            Execute.Assertion
+                .ForCondition(Subject.CompareTo(expected) == Equal)
+                .BecauseOf(because, becauseArgs)
+                .FailWith("Expected {context:object} {0} to be be ranked as equal to {1}{reason}.", Subject, expected);
```
Comment (jnyrup) [inline Src/FluentAssertions/Numeric/ComparableTypeAssertions.cs]: `be be`

**Comment by dennisdoomen** [comment]
Now the unit test is failing on the changed message.

**Pull request merged**